### PR TITLE
Realm event notification

### PR DESCRIPTION
**Pull request by dmlloyd**

### CODE REVIEW

**Comment by dmlloyd** [comment]
This might need a rebase after #287 so maybe merge that one first.

File: src/main/java/org/wildfly/security/auth/server/event/RealmIdentitySuccessfulAuthorizationEvent.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.wildfly.security.authz.AuthorizationIdentity;
+
+/**
+ * An event indicating that one identity attempted to authorize as another identity.
+ *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+public final class RealmIdentitySuccessfulAuthorizationEvent extends RealmIdentityAuthorizationEvent {
+    private final AuthorizationIdentity newAuthorizationIdentity;
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/event/RealmIdentitySuccessfulAuthorizationEvent.java]: On reflection, I don't like this.  This potentially allows an authorization identity from another realm to get passed to the target realm.  I think I'll change this to be more superficial, like old/new Principal or something.

File: src/main/java/org/wildfly/security/auth/server/event/RealmSuccessfulAuthenticationEvent.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.authz.AuthorizationIdentity;
+
+/**
+ * A realm event signifying authentication success.
+ *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+public final class RealmSuccessfulAuthenticationEvent extends RealmAuthenticationEvent {
+    private final AuthorizationIdentity authorizationIdentity;
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/event/RealmSuccessfulAuthenticationEvent.java]: In order to implement policies like storing rehashed passwords and password transformation, I think I should add the credential name used for authentication, as well as the credential itself, and clear password guess if one is available.

File: src/main/java/org/wildfly/security/auth/server/event/RealmAuthenticationEvent.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+package org.wildfly.security.auth.server.event;
+
+import org.wildfly.security.auth.server.RealmIdentity;
+
+/**
+ * A realm authentication event.  The realm identity may be destroyed at some point after the event is handled.
+ *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+public abstract class RealmAuthenticationEvent extends RealmEvent {
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/event/RealmAuthenticationEvent.java]: Need a RealmAbandonedAuthenticationEvent too...

File: src/main/java/org/wildfly/security/auth/server/event/RealmDefiniteOutcomeAuthenticationEvent.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public Object getCredential() {
+        return credential;
+    }
+
+    /**
+     * Get the actual credential guess used.
+     *
+     * @return the actual credential guess used, or {@code null} if there was no guess, it is not known, or no credential was used
+     */
+    public Object getEvidence() {
```
Comment (darranl) [inline src/main/java/org/wildfly/security/auth/server/event/RealmDefiniteOutcomeAuthenticationEvent.java]: Should the return type be Evidence?

File: src/main/java/org/wildfly/security/auth/server/event/RealmDefiniteOutcomeAuthenticationEvent.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    public String getCredentialName() {
+        return credentialName;
+    }
+
+    /**
+     * Get the actual credential used.
+     *
+     * @return the actual credential used, or {@code null} if it is not known or none was used
+     */
+    public Object getCredential() {
```
Comment (darranl) [inline src/main/java/org/wildfly/security/auth/server/event/RealmDefiniteOutcomeAuthenticationEvent.java]: Should the return type be Credential?

**Comment by darranl** [comment]
Issue wise this is most related to https://issues.jboss.org/browse/ELY-35

**Comment by dmlloyd** [comment]
That's one downside to having PRs open for a while. :)

**Comment by dmlloyd** [comment]
Updated to use Credential/Evidence APIs.  Rebased because the commit history was becoming unnecessarily convoluted.

File: src/main/java/org/wildfly/security/auth/server/SecurityRealm.java [hunk 6]
```diff
@@ -66,6 +70,33 @@ default SupportLevel getEvidenceVerifySupport(String credentialName) throws Real
...
     /**
+     * Handle a realm event.  These events allow the realm to act upon occurrences that are relevant to policy of
+     * the realm; for example, the realm may choose to increase password iteration count on authentication success,
+     * or change the salt of a password after a certain number of authentications.
+     * <p>
+     * The default implementation does nothing.
+     *
+     * @param event the realm event
+     */
+    default void handleRealmEvent(RealmEvent event) {}
```
Comment (darranl) [inline src/main/java/org/wildfly/security/auth/server/SecurityRealm.java]: Before we merge all of our realm implementations that wrap other realm implementations e.g. aggregate / privileged will need to have this method added so it passes through.

**Pull request merged**